File: backend/src/models/tournament.py
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from src.models.arsenal import BowlingBall

Strategy = Literal["conservative", "versatile", "aggressive", "defensive"]
LineupRole = Literal["primary", "secondary", "tertiary", "spare"]


class LineupCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    tournament_name: Optional[str] = Field(default=None, max_length=200)
    pattern_id: Optional[UUID] = None
    strategy: Optional[Strategy] = None
    notes: Optional[str] = None


class LineupBallAddRequest(BaseModel):
    user_arsenal_id: UUID
    role: LineupRole
    notes: Optional[str] = None


class LineupBall(BaseModel):
    id: UUID
    user_arsenal_id: UUID
    role: LineupRole
    order_index: int
    notes: Optional[str]
    ball: BowlingBall


class TournamentLineup(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    tournament_name: Optional[str]
    pattern_id: Optional[UUID]
    strategy: Optional[str]
    notes: Optional[str]
    created_at: datetime


class TournamentLineupDetail(TournamentLineup):
    balls: list[LineupBall] = []


class LineupRecommendation(BaseModel):
    user_arsenal_id: UUID
    role: LineupRole
    reasoning: str
    ball: BowlingBall

File: frontend/src/types/tournament.ts
import type { BowlingBall } from './arsenal';

export type Strategy = 'conservative' | 'versatile' | 'aggressive' | 'defensive';
export type LineupRole = 'primary' | 'secondary' | 'tertiary' | 'spare';

export type LineupCreateInput = {
  name: string;
  tournament_name?: string;
  pattern_id?: string;
  strategy?: Strategy;
  notes?: string;
};

export type LineupBallAddInput = {
  user_arsenal_id: string;
  role: LineupRole;
  notes?: string;
};

export type LineupBall = {
  id: string;
  user_arsenal_id: string;
  role: LineupRole;
  order_index: number;
  notes: string | null;
  ball: BowlingBall;
};

export type TournamentLineup = {
  id: string;
  user_id: string;
  name: string;
  tournament_name: string | null;
  pattern_id: string | null;
  strategy: string | null;
  notes: string | null;
  created_at: string;
};

export type TournamentLineupDetail = TournamentLineup & {
  balls: LineupBall[];
};

export type LineupRecommendation = {
  user_arsenal_id: string;
  role: LineupRole;
  reasoning: string;
  ball: BowlingBall;
};
